### PR TITLE
Fixing bilateral filter

### LINKED ISSUE #3995: Bilateral filter is broken (going back to at least 0.14.x)
Current gallery example:
https://scikit-image.org/docs/stable/auto_examples/filters/plot_denoise.html

0.13.x gallery example:
https://scikit-image.org/docs/0.13.x/auto_examples/filters/plot_denoise.html

Same code.  Not subtle.

This may have been introduced in #3937, though it could have been earlier; the gallery example from the most current 0.14.x is also affected.  I'm rather surprised it passes tests.

### PR DESCRIPTION
**Pull request by alexdesiqueira**
## Description

Bug fix. Closes #3995 .

## Checklist

- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

File: skimage/restoration/_denoise.py [hunk 1]
```diff
@@ -10,39 +9,89 @@
...
+    dtype : data type object
+        The type and size of the data to be returned.
 
+    Returns:
+    --------
+    color_lut : ndarray
+        Lookup table for the color distance sigma.
+    """
+    values = np.linspace(0, max_value, bins, endpoint=False)
+    return _gaussian_filter(values, sigma**2, dtype=dtype)
```
Comment (jni) [inline skimage/restoration/_denoise.py]: I don't think you should rename `_gaussian_weights`. This is not a filter, it's a weighting.

**Comment by jni** [comment]
Wow, nice fix! Does it actually fix the gallery example? Also, ideally there would be a test that fails with the old code but passes with the new code, since our existing tests did not catch the break.

**Comment by alexandrejaguar** [comment]
@jni it does; the effect was caused by the window size + a condition on the bin number. I wrote a new test to check the padding.

File: skimage/restoration/_denoise.py [hunk 2]
```diff
@@ -10,39 +9,89 @@
...
 
 
 def _compute_color_lut(bins, sigma, max_value, *, dtype=float):
-    values = np.linspace(0, max_value / bins, bins, endpoint=False)
-    color_lut = _gaussian_weight(sigma**2, values, dtype=dtype)
-    return color_lut
+    """Helping function. Define a lookup table containing Gaussian filter
+    values using the color distance sigma.
+
+    Parameters:
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: nit-picky comments ON.
```suggestion
    Parameters
```
And below... Don't forget to shorten the dashes.

File: skimage/restoration/_denoise.py [hunk 3]
```diff
@@ -10,39 +9,89 @@
...
+    sigma : float
+        Standard deviation for grayvalue/color distance (radiometric
+        similarity). A larger value results in averaging of pixels with larger
+        radiometric differences. Note, that the image will be converted using
+        the `img_as_float` function and thus the standard deviation is in
+        respect to the range ``[0, 1]``. If the value is ``None`` the standard
+        deviation of the ``image`` will be used.
+    max_value : float
+        Maximum value of the input image.
+    dtype : data type object
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]:
```suggestion
    dtype : data type object, optional
```
(and elsewhere).

File: skimage/restoration/_denoise.py [hunk 4]
```diff
@@ -79,9 +128,24 @@ def denoise_bilateral(image, win_size=None, sigma_color=None, sigma_spatial=1,
...
+    Radiometric similarity is measured by the Gaussian function of the
+    Euclidean distance between two color values and a certain standard
+    deviation (`sigma_color`).
+
     References
     ----------
-    .. [1] https://users.soe.ucsc.edu/~manduchi/Papers/ICCV98.pdf
+    .. [1] C. Tomasi and R. Manduchi. "Bilateral Filtering for Gray and Color
+           Images." IEEE International Conference on Computer Vision (1998)
+           839-846. :DOI: `10.1109/ICCV.1998.710815`
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]:
```suggestion
           839-846. :DOI:`10.1109/ICCV.1998.710815`
```
no space.

**Comment by jni** [comment]
Oh sorry: I thought @sciunto had also approved. I guess you'll have to wait for his approval before merging. =P

File: skimage/restoration/_denoise.py (hunk 2 above)
Comment (alexandrejaguar) [reply skimage/restoration/_denoise.py]: I mixed the style of our docs with something I was typing here these days :sweat_smile: Sorry about that!

**Comment by alexandrejaguar** [comment]
:man_facepalming:

**Comment by hmaarrfk** [comment]
No. You are experiencing

https://github.com/scikit-image/scikit-image/pull/4082

Feel free to pull that in.

**Comment by JDWarner** [comment]
Thanks for the fix @alexdesiqueira! All of @sciunto's recommendations are now included.  Merging.

**Pull request merged**